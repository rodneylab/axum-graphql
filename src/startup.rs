use std::future::ready;

use axum::{extract::Extension, middleware, routing::get, serve::Serve, Router};
use metrics_exporter_prometheus::PrometheusHandle;
use sqlx::SqlitePool;
use tokio::signal;
use tower_http::{compression::CompressionLayer, services::ServeDir, timeout::TimeoutLayer};

use crate::{
    database::run_migrations,
    model::get_schema,
    observability::metrics::track as track_metrics,
    routes::{graphql_handler, graphql_playground, health},
};

pub struct ApplicationRouters {
    pub main_router: Router,
    pub metrics_router: Router,
}

impl ApplicationRouters {
    /// Build the main app and metrics app routers.  These routers can be used in unit tests.
    ///
    /// # Errors
    /// Returns an error if the database is not reachable
    ///
    /// This function will return an error if .
    pub async fn build(
        database_url: &str,
        recorder_handle: PrometheusHandle,
    ) -> Result<Self, std::io::Error> {
        Ok(Self {
            main_router: main_router(database_url).await,
            metrics_router: metrics_router(recorder_handle),
        })
    }
}

/// Listen for and handle shutdown signals.
///
/// # Panics
///
/// Panics if unable to install Ctrl +C handler.
pub async fn shutdown_signal() {
    let ctrl_c = async {
        signal::ctrl_c()
            .await
            .expect("failed to install Ctrl+C handler");
    };

    #[cfg(unix)]
    let terminate = async {
        signal::unix::signal(signal::unix::SignalKind::terminate())
            .expect("failed to install signal handler")
            .recv()
            .await;
    };

    #[cfg(not(unix))]
    let terminate = std::future::pending::<()>();

    tokio::select! {
        () = ctrl_c => {
            tracing::info!("Ctrl-C registered");
            opentelemetry::global::shutdown_tracer_provider();
        },
        () = terminate => {
            tracing::info!("Terminate registered");
            opentelemetry::global::shutdown_tracer_provider();
        },
    }
}

pub struct Application {
    pub main_server: Serve<Router, Router>,
    pub main_server_port: u16,
    pub metrics_server: Serve<Router, Router>,
    pub metrics_server_port: u16,
}

impl Application {
    /// Build an axum main app and a metrics app.
    ///
    /// # Panics
    /// Panics if the database is not reachable, if the main app or merics port is already in use.
    ///
    /// # Errors
    ///
    /// This function will return an error if the metric port is already in use.
    pub async fn build(
        database_url: &str,
        recorder_handle: PrometheusHandle,
        (main_listener_ip, main_listener_port): (&str, u16),
        (metrics_listener_ip, metrics_listener_port): (&str, u16),
    ) -> Result<Self, std::io::Error> {
        let ApplicationRouters {
            main_router,
            metrics_router,
        } = ApplicationRouters::build(database_url, recorder_handle)
            .await
            .expect("database should be reachable");

        let main_listener =
            tokio::net::TcpListener::bind(format!("{main_listener_ip}:{main_listener_port}"))
                .await
                .unwrap_or_else(|_| {
                    panic!("`{main_listener_ip}:{main_listener_port}` should not already be in use")
                });
        let main_server_port = main_listener.local_addr().unwrap().port();
        tracing::info!(
            "Main app service listening on {}",
            main_listener.local_addr().unwrap()
        );

        let metrics_listener =
            tokio::net::TcpListener::bind(format!("{metrics_listener_ip}:{metrics_listener_port}"))
                .await?;
        let metrics_server_port = metrics_listener.local_addr().unwrap().port();

        tracing::info!(
            "Metrics service listening on {}",
            metrics_listener.local_addr().unwrap()
        );

        Ok(Self {
            main_server: axum::serve(main_listener, main_router),
            main_server_port,
            metrics_server: axum::serve(metrics_listener, metrics_router),
            metrics_server_port,
        })
    }

    /// Run both apps.  Can be used in tests and when running the app in production.
    ///
    /// # Errors
    ///
    /// This function will return an error if the axum main server or metrics server returned an
    /// error.
    pub async fn run_until_stopped(self) -> Result<(), std::io::Error> {
        let (main_server, metrics_server) = tokio::join!(
            self.main_server.with_graceful_shutdown(shutdown_signal()),
            self.metrics_server
                .with_graceful_shutdown(shutdown_signal())
        );
        main_server?;
        metrics_server?;

        Ok(())
    }
}

/// Create the main app axum router.
///
/// # Panics
/// Panics when not able to reach the database.
///
/// Panics if .
pub async fn main_router(database_url: &str) -> Router {
    tracing::info!("Main app service starting");

    let db_pool = SqlitePool::connect(database_url)
        .await
        .expect("SQLite database should be reachable");
    run_migrations(&db_pool).await;

    let schema = get_schema(db_pool);

    Router::new()
        .route("/", get(graphql_playground).post(graphql_handler))
        .route("/health", get(health))
        // serve GraphQL Playground CDN assets locally
        .nest_service("/assets", ServeDir::new("public"))
        .layer(CompressionLayer::new())
        .layer((
            middleware::from_fn(track_metrics),
            TimeoutLayer::new(std::time::Duration::from_secs(10)),
        ))
        .layer(Extension(schema))
}

pub fn metrics_router(recorder_handle: PrometheusHandle) -> Router {
    Router::new().route("/metrics", get(move || ready(recorder_handle.render())))
}
