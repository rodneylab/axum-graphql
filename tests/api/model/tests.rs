use std::{path::Path, str};

use axum::{
    body::Body,
    http::{Method, Request, StatusCode, header},
};
use axum_graphql::startup::ApplicationRouter;
use futures::executor::block_on;
use http_body_util::BodyExt;
use serde_json::{Value, json};
use tower::{Service, ServiceExt};

use crate::helpers::TestApp;

mod helpers {
    use axum::{
        Router,
        body::Body,
        http::{Method, Request, header},
    };
    use http_body_util::BodyExt;
    use serde_json::{Value, json};
    use tower::{Service, ServiceExt};

    pub async fn create_draft(app: &mut Router, title: &str, body: &str) -> i64 {
        let create_draft_json_request_body: Value = json!({
            "operationName":"CreateDraftMutation",
            "variables":{},
            "query": format!(r#"mutation CreateDraftMutation {{
                 createDraft(title: "{title}", body: "{body}") {{
                     id
                     title
                 }}
            }}"#),
        });

        let request = Request::builder()
            .method(Method::POST)
            .uri("/")
            .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
            .body(Body::from(create_draft_json_request_body.to_string()))
            .unwrap();
        let response = ServiceExt::<Request<Body>>::ready(app)
            .await
            .unwrap()
            .call(request)
            .await
            .unwrap();
        let body = response.into_body().collect().await.unwrap().to_bytes();
        let json_data: Value = serde_json::from_slice(&body).unwrap();
        let id = &json_data["data"]["createDraft"]["id"];

        id.as_i64().unwrap()
    }

    pub async fn publish_draft(app: &mut Router, id: i64) {
        let publish_draft_json_request_body: Value = json!({
            "operationName":"PublishMutation",
            "variables":{},
            "query": format!("mutation PublishMutation {{
  publish(id: {id}) {{
    __typename
    ... on PublishSuccessResponse {{
      post {{
        id
        published
      }}
    }}
    ... on PublishErrorResponse {{
      error {{
        field
        message
        received
      }}
    }}
  }}
}}"),
        });

        let request = Request::builder()
            .method(Method::POST)
            .uri("/")
            .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
            .body(Body::from(publish_draft_json_request_body.to_string()))
            .unwrap();
        let _response = ServiceExt::<Request<Body>>::ready(app)
            .await
            .unwrap()
            .call(request)
            .await
            .unwrap();
    }
}

async fn snapshot_graqphql_query_async<P: AsRef<Path>>(path: P) {
    // arrange
    let ApplicationRouter { router } = TestApp::spawn_routers().await;
    let json_request_body: Value = serde_json::from_slice(
        &std::fs::read(&path).expect("file should exist and have read permissions set"),
    )
    .unwrap_or_else(|_| {
        panic!(
            "File `{}` should contain valid JSON",
            path.as_ref().display()
        )
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    let body_bytes = response.into_body().collect().await.unwrap().to_bytes();
    let body = str::from_utf8(&body_bytes).unwrap();
    let body_json: Value = serde_json::from_str(body).unwrap();
    insta::assert_json_snapshot!(body_json, {".extensions.traceId" => "[traceId]"});
    // assert_eq!(
    //     body_json["extensions"]["traceId"],
    //     "00000000000000000000000000000000"
    // );
}

#[tokio::test]
async fn snapshot_graphql_queries() {
    insta::glob!("snapshot_inputs/*.json", |path| {
        block_on(snapshot_graqphql_query_async(path));
    });
}

#[tokio::test]
async fn graphql_endpoint_responds_to_hello_query() {
    // arrange
    let ApplicationRouter { router } = TestApp::spawn_routers().await;
    let json_request_body: Value = json!({
        "operationName":"HelloQuery",
        "variables":{},
        "query":"query HelloQuery { hello }"
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "hello": "Hello everybody!" },
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn graphql_endpoint_responds_to_drafts_query() {
    // arrange
    let ApplicationRouter { mut router } = TestApp::spawn_routers().await;
    let drafts_json_request_body: Value = json!({
        "operationName":"DraftsQuery",
        "variables":{},
        "query":"query DraftsQuery { drafts { id title} }"
    });

    // act
    let request = Request::builder()
        .method(Method::POST)
        .uri("/")
        .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
        .body(Body::from(drafts_json_request_body.to_string()))
        .unwrap();
    let response = ServiceExt::<Request<Body>>::ready(&mut router)
        .await
        .unwrap()
        .call(request)
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "drafts": [] },
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );

    // arrange
    let create_draft_json_request_body: Value = json!({
        "operationName":"CreateDraftMutation",
        "variables":{},
        "query": r#"mutation CreateDraftMutation {
                 createDraft(title: "Draft title", body: "Draft body text") {
                     id
                     title
                 }
            }"#,
    });

    // act
    let request = Request::builder()
        .method(Method::POST)
        .uri("/")
        .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
        .body(Body::from(create_draft_json_request_body.to_string()))
        .unwrap();
    let response = ServiceExt::<Request<Body>>::ready(&mut router)
        .await
        .unwrap()
        .call(request)
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "createDraft": { "id": 1, "title": "Draft title" } },
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );

    // act
    let request = Request::builder()
        .method(Method::POST)
        .uri("/")
        .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
        .body(Body::from(drafts_json_request_body.to_string()))
        .unwrap();
    let response = ServiceExt::<Request<Body>>::ready(&mut router)
        .await
        .unwrap()
        .call(request)
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "drafts": [ { "id": 1, "title": "Draft title" } ] },
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn posts_return_empty_array_when_no_posts_exist() {
    // arrange
    let ApplicationRouter { router } = TestApp::spawn_routers().await;
    let posts_json_request_body: Value = json!({
        "operationName":"PostsQuery",
        "variables":{},
        "query":"query PostsQuery { posts { id title} }"
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(posts_json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "posts": [] },
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn posts_returns_existing_posts() {
    // arrange
    let ApplicationRouter { mut router } = TestApp::spawn_routers().await;
    let id_1 = helpers::create_draft(&mut router, "First Post Title", "First post body.").await;
    let _id_2 = helpers::create_draft(&mut router, "Second Post Title", "Second post body.").await;
    let id_3 = helpers::create_draft(&mut router, "Third Post Title", "Third post body.").await;
    helpers::publish_draft(&mut router, id_1).await;
    helpers::publish_draft(&mut router, id_3).await;
    let posts_json_request_body: Value = json!({
        "operationName":"PostsQuery",
        "variables":{},
        "query":"query PostsQuery { posts { id title} }"
    });

    // act
    let request = Request::builder()
        .method(Method::POST)
        .uri("/")
        .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
        .body(Body::from(posts_json_request_body.to_string()))
        .unwrap();
    let response = ServiceExt::<Request<Body>>::ready(&mut router)
        .await
        .unwrap()
        .call(request)
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "posts": [
                { "id": id_1, "title": "First Post Title" },
                { "id": id_3, "title": "Third Post Title" },
            ]},
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn publish_returns_user_error_for_invalid_id() {
    // arrange
    let ApplicationRouter { router } = TestApp::spawn_routers().await;
    let id = 9_999;
    let publish_draft_json_request_body: Value = json!({
        "operationName":"PublishMutation",
        "variables":{},
        "query": format!("mutation PublishMutation {{
  publish(id: {id}) {{
    __typename
    ... on PublishSuccessResponse {{
      post {{
        id
        published
      }}
    }}
    ... on PublishErrorResponse {{
      error {{
        field
        message
        received
      }}
    }}
  }}
}}"),
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(publish_draft_json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "publish": {
                "__typename": "PublishErrorResponse",
                "error": {
                    "field": "id",
                    "message": "Did not find draft post with id `9999`",
                    "received": "9999"
                }
            }},
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn publish_returns_user_expected_result_for_valid_input() {
    // arrange
    let ApplicationRouter { mut router } = TestApp::spawn_routers().await;
    let _id_1 = helpers::create_draft(&mut router, "First Post Title", "First post body.").await;
    let id_2 = helpers::create_draft(&mut router, "Second Post Title", "Second post body.").await;
    let _id_3 = helpers::create_draft(&mut router, "Third Post Title", "Third post body.").await;
    let publish_draft_json_request_body: Value = json!({
        "operationName":"PublishMutation",
        "variables":{},
        "query": format!("mutation PublishMutation {{
  publish(id: {id_2}) {{
    __typename
    ... on PublishSuccessResponse {{
      post {{
        id
        published
      }}
    }}
    ... on PublishErrorResponse {{
      error {{
        field
        message
        received
      }}
    }}
  }}
}}"),
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(publish_draft_json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "publish": {
                "__typename": "PublishSuccessResponse",
                "post": {
                    "id": id_2,
                    "published": true,
                }
            }},
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn delete_draft_returns_user_error_for_invalid_id() {
    // arrange
    let ApplicationRouter { router } = TestApp::spawn_routers().await;
    let id = 9_999;
    let delete_draft_json_request_body: Value = json!({
        "operationName":"DeleteDraftMutation",
        "variables":{},
        "query": format!("mutation DeleteDraftMutation {{
  deleteDraft(id: {id}) {{
    __typename
    ... on DeleteDraftSuccessResponse {{
      post {{
        id
        title
      }}
    }}
    ... on DeleteDraftErrorResponse {{
      error {{
        field
        message
        received
      }}
    }}
  }}
}}"),
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(delete_draft_json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "deleteDraft": {
                "__typename": "DeleteDraftErrorResponse",
                "error": {
                    "field": "id",
                    "message": "Did not find draft post with id `9999`",
                    "received": "9999"
                }
            }},
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}

#[tokio::test]
async fn delete_draft_returns_user_expected_result_for_valid_input() {
    // arrange
    let ApplicationRouter { mut router } = TestApp::spawn_routers().await;
    let _id_1 = helpers::create_draft(&mut router, "First Post Title", "First post body.").await;
    let id_2 = helpers::create_draft(&mut router, "Second Post Title", "Second post body.").await;
    let _id_3 = helpers::create_draft(&mut router, "Third Post Title", "Third post body.").await;
    let delete_draft_json_request_body: Value = json!({
        "operationName":"DeleteDraftMutation",
        "variables":{},
        "query": format!("mutation DeleteDraftMutation {{
  deleteDraft(id: {id_2}) {{
    __typename
    ... on DeleteDraftSuccessResponse {{
      post {{
        id
        title
      }}
    }}
    ... on DeleteDraftErrorResponse {{
      error {{
        field
        message
        received
      }}
    }}
  }}
}}"),
    });

    // act
    let response = router
        .oneshot(
            Request::builder()
                .method(Method::POST)
                .uri("/")
                .header(header::CONTENT_TYPE, mime::APPLICATION_JSON.as_ref())
                .body(Body::from(delete_draft_json_request_body.to_string()))
                .unwrap(),
        )
        .await
        .unwrap();

    // assert
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(
        body,
        json!({
            "data": { "deleteDraft": {
                "__typename": "DeleteDraftSuccessResponse",
                "post": {
                    "id": id_2,
                    "title": "Second Post Title"
                }
            }},
            "extensions": { "traceId": "00000000000000000000000000000000" }
        })
    );
}
