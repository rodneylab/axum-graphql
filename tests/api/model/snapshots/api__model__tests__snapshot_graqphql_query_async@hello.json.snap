---
source: tests/api/model/tests.rs
expression: body_json
input_file: tests/api/model/snapshot_inputs/hello.json
snapshot_kind: text
---
{
  "data": {
    "hello": "Hello everybody!"
  },
  "extensions": {
    "traceId": "[traceId]"
  }
}
